refactor: menu & tray

lemarier opened the pull request:
<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update
- [x] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)
<!--
If yes, please describe the impact and migration path for existing applications in an attached issue. Filing a PR with breaking changes that has not been discussed and approved by the maintainers in an issue will be immediately closed.
-->

- [x] Yes. Issue #___
- [ ] No


**The PR fulfills these requirements:**

- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

### Todo

- [x] Add new event `TrayEvent` to track left, right and double click on tray icon (<s>macOS</s>, <s>Windows</s>, Linux). 
- [X] Add `show_menu()` `hide_menu()` to `Window`
- [x]  Move `NativeImage` behind a `PlatformExt`
- [x] WIndows `MenuItem` variants similar to what we've done with Linux
- [x] Send coords (bounds) of the tray icon if possible within `TrayEvent` If possible (<s>macOS</s>, <s>Windows</s>, Linux)
- [x] Only forward `WM_COMMAND` messages that corresponds to a menu item as `MenuEvent` on windows
- [x] Big cleanup
- [ ]  <s>Add `showContextMenu(menu)` to tray</s> -- will do in another release as macOS seems complicated
- [ ] `set_icon` for Windows, Linux and macOS (custom image) -- could be done in another release

***

from @amrbashir 
```
gtk c++ docs says there is GtkImageMenuItem but it is deprecated so gtk-rs doesn't have that
```

Fix #20 
Fix #47
Fix #63 
Fix #64 
Fix #66 
Fix #67

lemarier committed:
wip menu builder refactoring

lemarier committed:
[ci skip] fmt

lemarier committed:
[ci skip] fix sub_menu (children)

lemarier committed:
allow hot swap of tray icons

lemarier committed:
[ci skip] cleanup

lemarier committed:
[ci skip] fix submenu

wusyong commented on src/menu.rs:
```diff
@@ -6,14 +6,57 @@ use std::{
+    (menu_id, item)
+  }
+
+  pub fn add_system_item(&mut self, item: SystemMenu, menu_type: MenuType) -> MenuItem {
```
I prefer to keep the name of `MenuItem`. Could related types and methods fall back to this accordingly?

wusyong commented on src/menu.rs:
```diff
@@ -6,14 +6,57 @@ use std::{
+    self.0.add_children(children.0, title, enabled);
+  }
+
+  pub fn add_item(
```
Ditto. So maybe rename this to `add_custom_item`?

wusyong commented on src/menu.rs:
```diff
@@ -6,14 +6,57 @@ use std::{
+pub struct MenuBuilder(pub(crate) PlatformMenuBuilder);
+
+impl MenuBuilder {
+  pub fn init() -> Self {
```
`MenuType` should be passed here instead imho.
Also, `new` is a more common constructor name in rust.

wusyong commented on src/menu.rs:
```diff
@@ -6,14 +6,57 @@ use std::{
+use crate::platform_impl::{MenuBuilder as PlatformMenuBuilder, MenuItem};
+
+#[derive(Debug, Clone)]
+pub struct MenuBuilder(pub(crate) PlatformMenuBuilder);
```
Looks like `Menu` isn't used anymore. Could you rename `MenuBuilder` to `Menu` to replace the previous one?

lemarier committed:
[ci skip] cleanup

lemarier committed:
add some comments

lemarier committed:
[ci skip] system tray cleanup

lemarier committed:
[ci skip] wrap system_menu into an option

lemarier committed:
wip windows

lemarier committed:
fix windows import

lemarier committed:
fix window handler

lemarier committed:
[ci skip] cleanup

lemarier committed:
add popup menu for windows in tray / context

lemarier committed:
fix `window_process` for windows, to track submenu click

lemarier committed:
make `subclass_proc` public

lemarier committed:
add todo comment

lemarier committed:
clean tray for windows

amrbashir committed:
implement `MenuItem` for windows

lemarier committed:
[ci skip] implement basic menu icons

lemarier committed:
Merge remote-tracking branch 'origin/dev' into feat/menu-refactor

lemarier committed:
WIP linux implementation

lemarier committed:
allocate `GtkMenuItem` to the item to allow editing

lemarier committed:
[ci skip] wip migrate

lemarier committed:
[ci skip] fix linux and macos migration

lemarier committed:
fix windows

lemarier committed:
complete `MenuItem` migration

lemarier committed:
[ci skip] add some documentation

lemarier committed:
add empty signature for tray when flag is off

lemarier committed:
rename `children` to `submenu`

amrbashir commented on src/menu.rs:
```diff
@@ -52,9 +52,9 @@ impl Menu {
   }
 
-  pub fn add_children(&mut self, children: Menu, title: &str, enabled: bool) {
+  pub fn add_submenu(&mut self, submenu: Menu, title: &str, enabled: bool) {
```
shouldn't this function signature follow the same as `MenuItem::Submenu` so it would be something like this `pub fn add_submenu(&mut self, title: &str, enabled: bool, submenu: Menu)`

lemarier replied:
Perfect thx

amrbashir commented on src/menu.rs:
```diff
@@ -178,43 +233,14 @@ pub enum MenuItem {
-  }
 
-  /// Assign keyboard shortcut to the menu action. Works only with `MenuItem::Custom`.
+  /// Represents a Submenu(label, enabled, sub menu)
```
should this be `Submenu(title, enabled, sub menu)` or we change the `title` argument in all of our menu functions to be `label`

lemarier committed:
replace submenu signature

lemarier committed:
fix doc

lemarier committed:
implement empty signature for ios & android

lemarier committed:
fix linux `set_icon` for tray

lemarier committed:
fix CI

lemarier committed:
fmt

lemarier committed:
fix system_tray feature

lemarier committed:
fix android & ios build

lemarier committed:
add `selected` to linux CustomMenuItem

amrbashir committed:
simplify windows `CustomMenuItem` impl

lemarier committed:
WIP macOS implementation of cursor click on tray icon

lemarier committed:
[ci skip] example focus window if already created

lemarier committed:
cleanup

amrbashir committed:
implement `enabled` menu item on linux

amrbashir committed:
toggle fullscreen on linux

amrbashir committed:
fix `enabled` on linux menu

amrbashir committed:
add `copy`,`paste`,`cut` menu items on windows

lemarier committed:
add macos example without dock/menu

lemarier committed:
Merge branch 'feat/menu-refactor' of github.com:tauri-apps/tao into feat/menu-refactor

lemarier committed:
Merge remote-tracking branch 'origin/dev' into feat/menu-refactor

amrbashir committed:
add `show_menu` & `hide_menu` to `Window`

amrbashir committed:
Merge branch 'dev' into feat/menu-refactor

amrbashir committed:
simplify `SendInput` call on windows for `copy/cut/paste` menu items

amrbashir committed:
add more menu item variants on windows

lemarier committed:
add `bounds` to `TrayEvent`

amrbashir committed:
fix system_tray on windows

amrbashir committed:
tray events on windows

lemarier committed:
[ci skip] fix example

lemarier committed:
Merge branch 'feat/menu-refactor' of github.com:tauri-apps/tao into feat/menu-refactor

lemarier committed:
replace `MenuIcon` with `NativeImage`

amrbashir committed:
fix `SystemTray::set_hicon` on windows

amrbashir committed:
remove `MenuIcon` on other platforms

lemarier committed:
[ci skip] cleanup

lemarier committed:
Merge branch 'feat/menu-refactor' of github.com:tauri-apps/tao into feat/menu-refactor

lemarier committed:
[ci skip] fix examples

lemarier committed:
fix linux build

amrbashir committed:
fix clippy

amrbashir committed:
fix fmt

amrbashir committed:
clean some windows stuff

amrbashir committed:
more cleanup

amrbashir committed:
add `position` to `TrayEvent`

lemarier committed:
macOS: add `position` and right/ctrl+left click event for tray

amrbashir commented on src/platform_impl/macos/system_tray.rs:
```diff
@@ -75,6 +81,13 @@ impl SystemTrayBuilder {
+        button,
+        sendActionOn: NSEventMask::NSLeftMouseDownMask
+          | NSEventMask::NSRightMouseDownMask
+          | NSEventMask::NSKeyDownMask
+          | NSEventMask::NSKeyDownMask
```
is this duplication necessary or accident ?

lemarier replied:
oooops! thanks

lemarier committed:
[ci skip] fix duplication `NSKeyDownMask`

lemarier committed:
WIP: clean api and add comments

lemarier committed:
[ci skip] add `platform_item` for linux

amrbashir commented on src/menu.rs:
```diff
@@ -8,11 +8,16 @@ use std::{
-// test with shortcuts, we can remove them as well?
+// Tray menu (Also known as context menu)
 pub struct Tray;
+// Menubar menu (It represent the Window menu for Windows and Linux)
```
Should these be 3 slashes instead of 2 ?

amrbashir commented on src/menu.rs:
```diff
@@ -32,67 +41,63 @@ impl Menubar {
-/// A menu item, bound to a pre-defined action or `Custom` emit an event. Note that status bar only
-/// supports `Custom` menu item variants. And on the menu bar, some platforms might not support some
-/// of the variants. Unsupported variant will be no-op on such platform.
+/// A menu item, bound to a pre-defined action or `Custom` emit an event. Note some platforms
+/// might not support some of the variants. Unsupported variant will be no-op on such platform.
```
Should we only just keep the pre-defined items that are possible on all platforms and remove the rest? it would be less confusing for users.

lemarier replied:
Let see what @wusyong have to say about that

lemarier committed:
linux remove `add_custom_item`

lemarier committed:
fmt

amrbashir reviewed (commented):
There is bunch of clippy warnings related to docs and some redundant clones, check them on the checks tab of the PR.

lemarier committed:
fix ci

lemarier committed:
remove `platform_item` from MenuItem::Custom

amrbashir committed:
fix windows

lemarier committed:
fix ios & android

lemarier committed:
allow using `MenuItem::Custom` struct directly

lemarier committed:
make clippy happy

lemarier committed:
better `CustomMenuItem` and `NativeImage` impl

lemarier committed:
fix ci

amrbashir committed:
cleanup

amrbashir committed:
forward `WM_COMMAND` only from menu items click

lemarier committed:
[ci skip] add platform specific for tray event

lemarier committed:
[ci skip] fix doc

lemarier committed:
Merge branch 'feat/menu-refactor' of github.com:tauri-apps/tao into feat/menu-refactor

lucasfernog commented:
Don't forget to add a change file (and it must be a `minor` bump) :)

amrbashir committed:
remove `WININFO_STASH` on windows

amrbashir committed:
swap `lazy_static` with `once_cell`

wusyong commented on src/menu.rs:
```diff
@@ -6,36 +6,98 @@ use std::{
+
+  /// Shortcut to add a submenu to this `Menu`.
+  ///
+  /// This function is equivalent to [`Menu::add_item(self, MenuItem::Submenu(title, enabled, submenu))`].
```
This won't get the link
```suggestion
  /// This function is equivalent to `Menu::add_item(self, MenuItem::Submenu(title, enabled, submenu))`.
```

wusyong commented on src/menu.rs:
```diff
@@ -178,45 +240,98 @@ pub enum MenuItem {
-  /// - **Windows / Android / iOS:** Unsupported
+  /// - **Android / iOS:** Unsupported
   ///
+  Submenu(String, bool, MenuPlatform),
```
Enum variant can also be a struct. What do you think about change to struct instead?

amrbashir committed:
Revert "swap `lazy_static` with `once_cell`"

This reverts commit 7b39058af2b03c513714a7f5a7cf86f46bd35a6e.

lemarier committed:
Merge remote-tracking branch 'origin/dev' into feat/menu-refactor

lemarier committed:
implement deref and rename to `MenuContext` and `MenuBar`

lemarier committed:
use a struct for `Submenu`

lemarier committed:
rename to `ContextMenu`

lemarier committed:
align `MenuType` with `MenuBar` and `ContextMenu`

lemarier committed:
fix linux build

lemarier committed:
make `Menu` private to the crate

lemarier committed:
WIP simplify `add_item` and add `add_native_item`

lemarier committed:
remove `Custom` MenuItem for linux

lemarier committed:
remove `Custom` MenuItem for windows

lemarier committed:
fix examples

lemarier committed:
fix iOS and Android signature

lemarier committed:
fmt

amrbashir committed:
`TrayEvent` is now dpi-aware on windows

amrbashir committed:
fix `SystemTray::remove()` on windows

lemarier committed:
add changelog

lemarier committed:
Merge branch 'feat/menu-refactor' of github.com:tauri-apps/tao into feat/menu-refactor

lemarier committed:
refactor `system_tray` impl

lemarier committed:
fmt

lemarier committed:
fix clippy warning

lemarier committed:
make clippy happy

lemarier committed:
[ci skip] fix doc

lemarier committed:
[ci skip] add more docs to events

amrbashir commented on .changes/menu-refactor.md:
```diff
@@ -0,0 +1,11 @@
+
+**Breaking change**: New menu/tray API.
+
+System tray now expose `set_icon()` to update the tray icon after initialization. The `system_tray::SystemTrayBuilder` has been moved to the root of the package as a module and available on Windows, Linux and macOS, only when the `tray` feature is enabled. Windows expose a `remove()` function available with `SystemTrayExtWindows`.
```
we should probably mention the `remove()` function somewhere in the docs and warn users that if they don't call it on windows, the icon won't be removed automatically.

lemarier replied:
let me know if it make sense;
https://github.com/tauri-apps/tao/pull/77/commits/49fce6a87927b9bf40ec94a6b293b4cf7aeaf246#diff-de45fd7bc91c961341f8b97439d08aaaefe616342ef4fa5ab235fda4b938afd2R50

amrbashir replied:
perfect!

lemarier committed:
rename `ClickType` to `TrayEvent` and add more docs on `SystemTray`

lemarier committed:
add `quit` example in tray menu with `SystemTrayExtWindows`

lemarier committed:
[ci skip] update doc

lemarier committed:
add error handling in windows `system_tray`

lemarier committed:
fix windows build

lemarier committed:
fix windows (part 2 ;/)

lucasfernog committed:
feat: use struct on `fn add_item`

lemarier committed:
create windows examples — left click open or focus window

lemarier committed:
Merge branch 'feat/menu-refactor' of github.com:tauri-apps/tao into feat/menu-refactor

lucasfernog committed:
feat: expose add_native_item on ContextMenu

lucasfernog committed:
fix: macos, android, ios impl

lemarier committed:
[ci skip] add doc for linux

Commit:
Introduce CheckMenuItem on linux

Commit:
Implement set_selected on linux

Commit:
Revert example change

wusyong commented on src/menu.rs:
```diff
@@ -175,46 +283,46 @@ pub enum MenuItem {
+/// Custom menu item, when clicked an event is emitted in the EventLoop.
+/// You can modify the item after it's creation.
+#[derive(Debug, Clone)]
+pub struct CustomMenuItemHandle(pub CustomMenuItemPlatform);
```
I think we should rename `CustomMenuItem` to `CustomMenuItemAttributes` or `MenuItemAttributes`, and then rename `CustomMenuItemHandle` to `CustomMenuItem` to follow the convention of this lib.

lemarier replied:
I'm not sure about this;

```rust
submenu.add_item(MenuItemAttributes::new("Open new window", None, true, false));
```

or

```rust
submenu.add_item(CustomMenuItemAttributes::new("Open new window", None, true, false));
```

look a bit weird from my end

lemarier replied:
What's about `MenuItemBuilder`

```rust
tray_menu.add_item(MenuItem::new("Focus window").with_selected(true));
```

rename `MenuItem` for `NativeItem` as it's used by `fn add_native_item`

wusyong commented on src/menu.rs:
```diff
@@ -6,37 +6,145 @@ use std::{
+  title: &'a str,
+  keyboard_accelerator: Option<&'a str>,
+  enabled: bool,
+  selected: bool,
```
This field could be `Option<bool>` to represent if the item should be a select item or not, and then select it or not with `Some(true)` and `Some(false)`.

lemarier replied:
i think it’s fair, im not against

lemarier replied:
What's about `MenuItemBuilder`

```rust
tray_menu.add_item(MenuItem::new("Focus window").with_selected(true));
```

rename `MenuItem` for `NativeItem` as it's used by `fn add_native_item`

wusyong replied:
Sounds good

lemarier committed:
- rename `CustomMenuItem` to `MenuItemAttributes`
- rename `CustomMenuItemHandle` to `CustomMenuItem`
- add `with_accelerators`, `with_enabled` and `with_selected` to `MenuItemAttributes`

lemarier committed:
[ci skip] clean example

lemarier committed:
[ci skip] fix example `custom_menu`

lemarier committed:
clean documentation and add some examples

lemarier committed:
fix `MenuItem` uppercase

lemarier committed:
fmt

lemarier committed:
add note for linux `system_tray_no_menu` and fix example

lemarier committed:
fix path linux example

lemarier committed:
add `quit` to linux example

lemarier committed:
fmt

lemarier committed:
fix menu_id

lemarier committed:
remove unwanted clone to id’s

lemarier committed:
some works on doc, specially `set_skip_taskbar`
